mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp

```diff
@@ -127,6 +127,46 @@ findAllocaInsertPoint(llvm::IRBuilderBase &builder,
       &funcEntryBlock, funcEntryBlock.getFirstInsertionPt());
 }
 
+static bool isOpAllowedToBeLowered(Operation *opInst,
+                                   llvm::OpenMPIRBuilder *ompBuilder) {
+  if (!opInst)
+    return false;
+  // omp.target operation can be lowered for host and device MLIR
+  if (isa<omp::TargetOp>(opInst))
+    return true;
+
+  // OpenMP operations inside omp.target can be lowered for host and device MLIR
+  if (opInst->getParentOfType<omp::TargetOp>())
+    return true;
+
+  // TODO: Add support for test case:
+  // omp.parallel { //host pragma
+  //   omp.target { }
+  // }
+  bool hasTargetRegion =
+      opInst->walk([](omp::TargetOp) { return WalkResult::interrupt(); })
+          .wasInterrupted();
+  if (hasTargetRegion)
+    opInst->emitError("Target region inside other pragma is not yet supported");
```

Comment:
Question for @skatrak  - What is the meaning of the a ``omp.target`` nested inside another ``omp.target``? Is it a NOP (I am guessing not since you refer to "reverse offloading")

Reply:
Thanks, I hope these kernels are big enough to make it worthwhile to do all the shuttling between host and device (or the cost of this shuttling is kept really low through zero-copy buffers and such things but still there can be such a thing as too much shuttling I think).